src/sources/file.rs

```diff
@@ -987,6 +987,7 @@ mod tests {
             let lines = extract_messages_string(received);
             assert_eq!(lines, vec!["zeroth line", "first line"]);
         }
+        sleep_500_millis().await;
```

Comment:
Weird, do I take it to mean that it takes this long for the checkpoint files to be fully flushed to disk from the first run?

Reply:
I think it means that the checkpoint writer, which executes asynchronously, *occasionally* finishes writing the file (and flushing any internal buffers) after the next server test starts and loads it. It doesn't take 500ms, those waits are probably overkill, but the function was already there so I re-used it.

Reply:
Hmm, it looks like `run_file_source` is shutting down the component before it returns though. I would imagine this includes stopping the checkpointing writer task?

Reply:
It should, yes, but only if the `wait_shutdown` condition is `true`, which I had missed. Turning it on should also resolve the problem, but doesn't appear to. I'll dig some more.

Reply:
I think this is happening because the checkpointer task occasionally fails to write the checkpoints a final time at shutdown. I'm running the test now with additional debug info, but that seems to have made it harder to trigger.

Reply:
From what I see in `file_server` the checkpointer and regular read future don't synchronize the shutdown, so the checkpointer can finish writing current state then shutdown while the read future is still reading some file, updating state and only then checking for shutdown condition. This seems like it fits

>  the checkpointer task occasionally fails to write the checkpoints a final time at shutdown